File: leetcode/cpp/22.cpp
#include "leetcode.h"

class Solution {
public:
	vector<string> generateParenthesis(int n)
	{
		vector<string> result;
		queue<tuple<string, int, int>> q;
		q.push(make_tuple("", 0, 0));
		while (!q.empty()) {
			auto t = q.front();
			q.pop();
			string cur = get<0>(t);
			int left = get<1>(t);
			int right = get<2>(t);
			if (left == n && right == n) {
				result.push_back(cur);
			} else {
				if (left < n) {
					q.push(make_tuple(cur+"(", left+1, right));
				}
				if (right < left) {
					q.push(make_tuple(cur+")", left, right+1));
				}
			}
		}

		return result;
	}
};

File: leetcode/golang/22.go
package leetcode

func generateParenthesisHelper(n int, left int, right int, cur string, result *[]string) {
	if left == n && right == n {
		*result = append(*result, cur)
	}

	if left < n {
		generateParenthesisHelper(n, left+1, right, cur+"(", result)
	}

	if right < left {
		generateParenthesisHelper(n, left, right+1, cur+")", result)
	}
}

func generateParenthesis(n int) []string {
	result := []string{}
	generateParenthesisHelper(n, 0, 0, "", &result)
	return result
}
